### PR TITLE
merge location and map matcher result observers

### LINKED ISSUE #4814: Merge LocationObserver with MapMatcherResultObserver
We currently offer 2 interfaces that provide redundant information. `LocationObserver` was the original implementation and provides only the information about `rawLocation`, `enhancedLocation`, and `keypoints`. During the course of `v1`, we introduced an additional `MapMatcherResultObserver` which next to providing the same information as `LocationObserver` also exposes matching probabilities, speed limits, and other matching information available in the free drive. We added a new interface for backward compatibility.

In my opinion, there's no need for these interfaces to live side-by-side. It only complicates the documentation and examples because while we can start with using `LocationObserver`, at some point for more advanced use-cases we need to migrate to the `MapMatcherResultObserver` anyway. There's no overhead in using `MapMatcherResultObserver` instead of `LocationObserver` so there's no reason for the SDK to maintain both.

I'd like to suggest alternative solutions:
- remove the `LocationObserver` and merge it with the `MapMatcherResultObserver` (potentially come up with a new name?)
- deprecate `LocationObserver` (do not remove it, do not change any names)

Since we're in the RC phase, we still have the opportunity to just go ahead and remove the `LocationObserver` but a deprecation can be a good path as well. I'd love to hear your thoughts @mapbox/navigation-android.

### PR DESCRIPTION
**Pull request by Zayankovsky**
### Description
Closes #4814. 

### Changelog

```
<changelog>Renamed `MapMatcherResult` to `LocationMatcherResult`</changelog>
<changelog>Merged `LocationObserver` and `MapMatcherResultObserver` interfaces into single `LocationObserver` interface</changelog>
```

### CODE REVIEW

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/AlternativeRouteActivity.kt [hunk 1]
```diff
@@ -107,27 +108,11 @@ class AlternativeRouteActivity : AppCompatActivity(), OnMapLongClickListener {
         initListeners()
     }
 
-    override fun onStart() {
-        super.onStart()
-        binding.mapView.onStart()
```
Comment (Guardiola31337) [inline qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/AlternativeRouteActivity.kt:112]: Is it safe to remove these? 👇

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/FeedbackActivity.kt [hunk 2]
```diff
@@ -163,30 +160,15 @@ class FeedbackActivity : AppCompatActivity() {
...
-
-    override fun onStop() {
-        super.onStop()
-        binding.mapView.onStop()
-    }
-
-    override fun onLowMemory() {
-        super.onLowMemory()
-        binding.mapView.onLowMemory()
-    }
```
Comment (Guardiola31337) [inline qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/FeedbackActivity.kt:179]: Same as above re: removing `mapView` lifecycle methods

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/LocationObserver.kt [hunk 3]
```diff
@@ -4,6 +4,8 @@ import android.location.Location
 
 /**
  * An interface which enables listening to location updates
+ *
+ * @see [LocationMatcherResult]
  */
 interface LocationObserver {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/LocationObserver.kt:10]: Do we still want to change the name of the `interface`? https://github.com/mapbox/mapbox-navigation-android/issues/4814#issuecomment-924601371 cc @LukasPaczos @abhishek1508

**Review by Guardiola31337 — COMMENTED** [review]
LGTM

Left some minor comments / questions.

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/AlternativeRouteActivity.kt (hunk 1 above)
Comment (Zayankovsky) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/AlternativeRouteActivity.kt:112]: Android Studio shows this warning:
Screenshot 2021-09-22 at 21 55 46
And I checked that Maps actually calls these methods themselves.

Comment (Guardiola31337) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/AlternativeRouteActivity.kt:112]: Thanks for confirming 🚀

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/LocationObserver.kt (hunk 3 above)
Comment (Zayankovsky) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/LocationObserver.kt:10]: I think `LocationMatcherResultObserver` would be a bit misleading, because apart from `LocationMatcherResult`s the interface allows to observe raw `Location`s.

**Review by Guardiola31337 — APPROVED** [review]

File: examples/src/main/java/com/mapbox/navigation/examples/core/IndependentRouteGenerationActivity.kt [hunk 4]
```diff
@@ -291,17 +289,10 @@ class IndependentRouteGenerationActivity : AppCompatActivity() {
...
-        binding.mapView.onStop()
     }
 
     override fun onDestroy() {
         super.onDestroy()
-        binding.mapView.onDestroy()
         mapboxNavigation.onDestroy()
     }
-
-    override fun onLowMemory() {
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/core/IndependentRouteGenerationActivity.kt:303]: This looks unrelated. What's the explanation here?

Comment (Guardiola31337) [reply examples/src/main/java/com/mapbox/navigation/examples/core/IndependentRouteGenerationActivity.kt:303]: https://github.com/mapbox/mapbox-navigation-android/pull/4880#discussion_r714232342

Comment (Zayankovsky) [reply examples/src/main/java/com/mapbox/navigation/examples/core/IndependentRouteGenerationActivity.kt:303]: > This looks unrelated. 

It actually is, sorry. I extracted those changes into a separate commit for a better history in git.
